tests/ci/pr_info.py

```diff
@@ -310,57 +310,73 @@ def has_changes_in_submodules(self):
         return False
 
     def can_skip_builds_and_use_version_from_master(self):
-        # TODO: See a broken loop
         if FORCE_TESTS_LABEL in self.labels:
             return False
 
         if self.changed_files is None or not self.changed_files:
             return False
 
-        for f in self.changed_files:
-            # TODO: this logic is broken, should be fixed before using
-            if (
-                not f.startswith("tests/queries")
-                or not f.startswith("tests/integration")
-                or not f.startswith("tests/performance")
-            ):
-                return False
-
-        return True
-
-    def can_skip_integration_tests(self):
-        # TODO: See a broken loop
+        return not any(
+            f.startswith("programs")
+            or f.startswith("src")
+            or f.startswith("base")
+            or f.startswith("cmake")
+            or f.startswith("rust")
+            or f == "CMakeLists.txt"
+            or f == "tests/ci/build_check.py"
+            for f in self.changed_files
+        )
+
+    def can_skip_integration_tests(self, versions: List[str]) -> bool:
         if FORCE_TESTS_LABEL in self.labels:
             return False
 
+        # If docker image(s) relevant to integration tests are updated
+        if any(self.sha in version for version in versions):
+            return False
+
         if self.changed_files is None or not self.changed_files:
             return False
 
-        for f in self.changed_files:
-            # TODO: this logic is broken, should be fixed before using
-            if not f.startswith("tests/queries") or not f.startswith(
-                "tests/performance"
-            ):
-                return False
+        if not self.can_skip_builds_and_use_version_from_master():
+            return False
 
-        return True
+        # Integration tests can be skipped if integration tests are not changed
+        return not any(
+            f.startswith("tests/integration/")
+            or f == "tests/ci/integration_test_check.py"
+            for f in self.changed_files
+        )
 
-    def can_skip_functional_tests(self):
-        # TODO: See a broken loop
+    def can_skip_functional_tests(
+        self, version: str, test_type: Literal["stateless", "stateful"]
+    ) -> bool:
         if FORCE_TESTS_LABEL in self.labels:
             return False
 
+        # If docker image(s) relevant to functional tests are updated
+        if self.sha not in version:
```

Comment:
It should be the other way around. If sha in the version, then the images are changed, so we run the tests

Reply:
right, sorry it is fine in integration check, will fix it